Update CONTRIBUTING.md (#617)
sentinel-hub/eo-learn: CONTRIBUTING.md

pip install -r requirements-dev.txt -r requirements-docs.txt
```

**Note:** to reduce later merge conflicts, always pull the latest version of the `develop` branch from the upstream eo-learn repository ([located here][dev-branch]) to your fork before starting the work on your PR.

### Contribute an `EOTask`

`EOTask`s are currently grouped into subpackages by scope, e.g. *core*, *IO*, *masks*, and so on. A list of implemented tasks can be found in the [documentation][existing-eo-tasks]. The following code snippet shows how to create your own `EOTask`:

```python
class FooTask(EOTask):
    def __init__(self, foo_params):
        self.foo_params = foo_params

    def execute(self, eopatch, *, execution_kwargs):
        # do what foo does on input eopatch and return it
        return eopatch
```

When creating a new task, bear in mind the following:

> * Tasks should be useful for multiple use-cases or solve a common problem i.e. other users would benefit from it.

Reviewer: ```suggestion
* Tasks should be useful for multiple use-cases or solve a common problem, i.e., other users should benefit from it.
```

* Use feature-parsers in order to support a wider range of possible input (and perhaps validate inputs). You can find them in `eolearn.core.utils.parsing`, or use `EOTask.parse_features` as a shortcut.
* If in doubt on whether a task is general enough to be of interest to the community, or you are not sure to which sub-package to contribute your task to, send us an email or open a [feature request](#feature-requests).

### Running formatters, linters, and tests

This section assumes you have installed all packages in `requirements-dev.txt`.

Most of the automated code-checking is packaged into [pre-commit hooks](https://pre-commit.com/). You can activate them by running `pre-commit install`. If you wish to check all code you can do so by running `pre-commit run --all-files`. This takes care of:
- auto-formatting the code using `black`, `isort`, and `autoflake`
- checking the code with `flake8`
- checking and formatting any Jupyter notebooks with `nbqa`
- various other helpful things (correcting line-endings etc.)

The code is also checked using `pylint` and `mypy`. Because of the project structure invoking these two checkers is a bit trickier:

> ```bash

Reviewer 2: suggestion: "... is a bit trickier and has to be performed on each module separately. If your contribution is in `core`, you could do the following:"

    def execute(self, eopatch, *, execution_kwargs):
        # do what foo does on input eopatch and return it
        return eopatch
```

When creating a new task, bear in mind the following:

* Tasks should be useful for multiple use-cases or solve a common problem, i.e., other users should benefit from it.
* In cases of large monolithic tasks you might want to try and split them into multiple ones, where each does just part of the computation. For instance a task for detecting foliage from S2 imagery could perhaps be split into a task for calculating NDVI and a separate task for thresholding the NDVI values. In many cases you'll find that some of these smaller tasks already exist in `eo-learn`.
* When multiple tasks share a large part of the implementation, consider extracting parts of it into functions or use inheritance to reduce boilerplate.
* Only hard-code parameters that shouldn't be adjusted by the users. Hard-coding things like feature names should be avoided.
* Use feature-parsers in order to support a wider range of possible input (and perhaps validate inputs). You can find them in `eolearn.core.utils.parsing`, or use `EOTask.parse_features` as a shortcut.

> * If in doubt on whether a task is general enough to be of interest to the community, or you are not sure to which sub-package to contribute your task to, send us an email or open a [feature request](#feature-requests).

Reviewer: ```suggestion
* If in doubt on whether a task is general enough to be of interest to the community, or you are not sure to which sub-package to contribute your task to, feel free to open up a [feature request](#feature-requests).
```

### Running formatters, linters, and tests

This section assumes you have installed all packages in `requirements-dev.txt`.

Most of the automated code-checking is packaged into [pre-commit hooks](https://pre-commit.com/). You can activate them by running `pre-commit install`. If you wish to check all code you can do so by running `pre-commit run --all-files`. This takes care of:
- auto-formatting the code using `black`, `isort`, and `autoflake`
- checking the code with `flake8`
- checking and formatting any Jupyter notebooks with `nbqa`
- various other helpful things (correcting line-endings etc.)

The code is also checked using `pylint` and `mypy`. Because of the project structure invoking these two checkers is a bit trickier and has to be performed on each module separately. If your contribution is in core, you could do the following:
```bash
# to check the 'core' subpackage
mypy core/eolearn/core
pylint core/eolearn/core
```
You can also use the utilities provided in the MAKEFILE to check the entire package with `make pylint` and `make mypy`. Due to the size of the codebase this might take a while.

> The last bit to check are the unit tests. We again reiterate that you should include unit-tests of your contributions. If you are not well versed with unit-testing you can ask us for help (either via e-mail or in the PR). To run the tests simply use the command `pytest` from the main folder. Since `eolearn.io` also test integration with SentinelHub you can either skip them with `pylint -m "not sh_integration"` or see the [examples/README.md](examples/README.md) for how to setup you SentinelHub account and local config for testing.

Reviewer: I feel like it would be better to remove statements where people should contact us via email for technical stuff 😬

Author: they never do anyway :P

Author: should we redirect them to issues instead?

> The last bit to check are the unit tests. We again reiterate that you should include unit-tests of your contributions. If you are not well versed with unit-testing you can ask us for help (either via e-mail or in the PR). To run the tests simply use the command `pytest` from the main folder. Since `eolearn.io` also test integration with SentinelHub you can either skip them with `pylint -m "not sh_integration"` or see the [examples/README.md](examples/README.md) for how to setup you SentinelHub account and local config for testing.

Reviewer: ```suggestion
The last bit to check are the unit tests. We again reiterate that you should include unit-tests of your contributions. If you are not well versed with unit-testing you can ask us for help in the pull-request or by issue a [ticket](https://github.com/sentinel-hub/eo-learn/issues) instead. To run the tests simply use the command `pytest` from the main folder. Since `eolearn.io` also test integration with SentinelHub you can either skip them with `pylint -m "not sh_integration"` or see the [examples/README.md](examples/README.md) for how to setup you SentinelHub account and local config for testing.
```

Author: this one has a typo i think, can you edit it?

Reviewer: right, changed.
